modules/publication-service-oaipmh/src/main/java/org/opencastproject/publication/oaipmh/OaiPmhPublicationServiceImpl.java

```diff
@@ -79,453 +80,648 @@
+      throw new IllegalArgumentException("OAI-PMH repository '" + repository + "' does not exist");
 
-  /**
-   * Callback for the OSGi declarative services configuration.
-   *
-   * @param distributionService
-   *          the distribution service
-   */
-  public void setStreamingDistributionService(StreamingDistributionService distributionService) {
-    this.streamingDistributionService = distributionService;
+    try {
+      return serviceRegistry.createJob(JOB_TYPE, Operation.Publish.toString(),
+              Arrays.asList(MediaPackageParser.getAsXml(mediaPackage), // 0
+                      repository, // 1
+                      StringUtils.join(downloadElementIds, SEPARATOR), // 2
+                      StringUtils.join(streamingElementIds, SEPARATOR), // 3
+                      Boolean.toString(checkAvailability))); // 4
+    } catch (ServiceRegistryException e) {
+      throw new PublicationException("Unable to create job", e);
+    }
   }
 
-  /**
-   * Callback for the OSGi declarative services configuration.
-   *
-   * @param persistence
-   *          the OAI-PMH persistence
-   */
-  public void setPersistence(OaiPmhDatabase persistence) {
-    this.persistence = persistence;
-  }
+  @Override
+  public Job retract(MediaPackage mediaPackage, String repository) throws PublicationException, NotFoundException {
+    if (mediaPackage == null)
+      throw new IllegalArgumentException("Media package must be specified");
 
-  /** OSGi DI. */
-  public void setOaiPmhServerInfo(OaiPmhServerInfo oaiPmhServerInfo) {
-    this.oaiPmhServerInfo = oaiPmhServerInfo;
-  }
+    if (StringUtils.isNotBlank(repository) && !oaiPmhServerInfo.hasRepo(repository))
+      throw new IllegalArgumentException("Not existant OAI-PMH repository " + repository);
 
-  /**
-   * Creates a new instance of the OAI-PMH publication service.
-   */
-  public OaiPmhPublicationServiceImpl() {
-    super(JOB_TYPE);
+    try {
+      return serviceRegistry.createJob(JOB_TYPE, Operation.Retract.toString(),
+              Arrays.asList(MediaPackageParser.getAsXml(mediaPackage), repository));
+    } catch (ServiceRegistryException e) {
+      throw new PublicationException("Unable to create job", e);
+    }
   }
 
   @Override
-  public Job publish(MediaPackage mediaPackage, String repository, Set<String> downloadIds, Set<String> streamingIds,
+  public Job updateMetadata(MediaPackage mediaPackage, String repository, Set<String> flavors, Set<String> tags,
           boolean checkAvailability) throws PublicationException, MediaPackageException {
     if (mediaPackage == null)
-      throw new MediaPackageException("Media package must be specified");
+      throw new IllegalArgumentException("Media package must be specified");
     if (StringUtils.isEmpty(repository))
-      throw new IllegalStateException("Repository must be specified");
+      throw new IllegalArgumentException("Repository must be specified");
+    if (!oaiPmhServerInfo.hasRepo(repository))
```

Comment:
Same code as in methods publish and retract. I think you really should factor the code into a method...